onnxruntime/core/framework/session_state.h

```diff
@@ -446,7 +448,8 @@ class SessionState {
   // and as this isn't considered performance critical currently it's not worth the maintenance overhead of adding one.
   // We do get an allocator from ExecutionFrame so this is looked up frequently, however there most likely aren't many
   // entries in the map
-  std::shared_ptr<std::map<OrtDevice, AllocatorPtr>> allocators_;
+  std::unique_ptr<AllocatorMap> allocators_unique_ptr_;
+  AllocatorMap* allocators_;
```

Comment:
Any reason you need this member?

Reply:
We cannot let SessionState object of subgraph hold unique pointer, right? because all SessionState objects of graph/subgraph will point the same allocators map, if SessionState object of subgraph holding unqiue pointer of allocators, the allocators map will be released when subgraph's session state is releasing but at the same time the parent SessionState is still alive who may still refer the allocators object

Reply:
I am little confused here. Based on your explanation, wouldn't it be better to use a shared ptr here? This way we create the allocators_shared_ptr when the SessionState of parent graph is created and all the SessionState objects (for subgraph) will just increment and decrement the ref count as they are created and destroyed. 

Irrespective of whether we stick with this or change to shared ptr, we need a comment here with some details here. Basically something like you added in the comment above.

Reply:
shared_ptrs are expensive. In this case we know that the subgraph SessionState is contained by the outer SessionState. Hence, there's no need for a shared_ptr.